### PR TITLE
[ExpansionPanelList] add materialGapSize property in ExpansionPanelList Widget

### LINKED ISSUE #118167: ExpansionPanelList should allow to set MaterialGap height
## Use case

I'd like to use an `ExpansionPanelList` without any gaps / spaces between their elements.

## Proposal

I'd like to define a `gapHeight` that will be passed to the `MaterialGap` widget that is placed between the expanded `ExpansionPanel` children of the `ExpansionPanelList`.

I can set the `expandedHeaderPadding`, but it will only influence the padding of the first header. Setting `elevation` does not solve the problem either, because the `ExpansionPanelList` does this in its build function: 

```dart
if (_isChildExpanded(index) && index != 0 && !_isChildExpanded(index - 1))
items.add(MaterialGap(key: _SaltedKey<BuildContext, int>(context, index * 2 - 1)));
```

`MaterialGap` accepts a `size` property in its constructor, so I wonder why there is no property on the `ExpandedPanelList` that allows me to set the `size` of those `MaterialGap`s.

## Illustration

I don't want these `ExpandedPanel`s to move when expanding. This is the closest I can get:

2023-01-09 5 39 35 AM

```dart
ExpansionPanelList(
      elevation: 0,
      dividerColor: Colors.transparent,
      expandedHeaderPadding: const EdgeInsets.symmetric(vertical: 0),
      expansionCallback: (int index, bool isExpanded) {
        setState(() {
          _expanded[index] = !isExpanded;
        });
      },
      children: panels,
    );
```

## Additional context

Please see: 

- https://stackoverflow.com/questions/60555041/is-there-a-way-to-get-rid-of-the-spacing-between-expansion-panels-when-expanded
- https://stackoverflow.com/questions/72822297/how-to-remove-space-between-expanded-expansionpanels-in-expansionpanellist
- https://stackoverflow.com/questions/60463999/flutterhow-to-remove-elevation-from-expansionpanellist

### PR DESCRIPTION
**Pull request by piedcipher**
Adds `materialGapSize` property to `ExpansionPanelList` widget.

| `materialGapSize: 0` | `materialGapSize: 16` (default) | `materialGapSize: 30` |
| --- | --- | --- |
| screenshot_1680450783 | screenshot_1680453525 | screenshot_1680450910 |
| screenshot_1680450785 | screenshot_1680453527 | screenshot_1680450912 |
| screenshot_1680450787 | screenshot_1680453530 | screenshot_1680450914 |
| screenshot_1680450789 | screenshot_1680453532 | screenshot_1680450916 |

Fixes: #118167

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (werainkhatri) [inline packages/flutter/lib/src/material/expansion_panel.dart:258]: seems like you haven't gone through the [Flutter Style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo). Do give it a read, specially [this](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc) section for documenting APIs.

generally, the format is like below. refer documentation of other members and methods for inspiration.

```dart
/// A short precise description.
///
/// More details disambiguating this API from similar ones, explaining edge cases, null or default behavior, etc.
///
/// See also section for related APIs.
```

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (werainkhatri) [inline packages/flutter/test/material/expansion_panel_test.dart:1623]: nit: `ExpansionPanelList.materialGapSize defaults to 16.0`

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (werainkhatri) [inline packages/flutter/test/material/expansion_panel_test.dart:1644]: i don't think this test holds much value, as you're just testing if the value you passed to the immediate argument is set correctly. it would be better to instead test if the effective gap b/w the items is correctly configured by `materialGapSize`.

after a short dive, i can see that `ExpansionPanelList` uses `MergeableMaterial` which adds `SizedBox` to a `ListBody` (which is like a `ListView`) for every `MaterialGap`. so you could test if the dimensions of the `SizedBox` are as per the value of `materialGapSize`.

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (werainkhatri) [inline packages/flutter/lib/src/material/expansion_panel.dart:174]: nit: `16.0`.

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (werainkhatri) [inline packages/flutter/lib/src/material/expansion_panel.dart:201]: nit: `16.0`.

**Review by werainkhatri — COMMENTED** [review]
the images are quite helpful, thanks for that!

left a few comments.

**Comment by piedcipher** [comment]
thanks a lot for the detailed review, I'll work on the review comments. 💙

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (piedcipher) [reply packages/flutter/lib/src/material/expansion_panel.dart]: done 👍✅

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (piedcipher) [reply packages/flutter/lib/src/material/expansion_panel.dart]: done 👍✅

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (piedcipher) [reply packages/flutter/test/material/expansion_panel_test.dart]: done 👍 o

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (piedcipher) [reply packages/flutter/lib/src/material/expansion_panel.dart:258]: Sure I'll go through this. My bad.

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (piedcipher) [reply packages/flutter/test/material/expansion_panel_test.dart:1644]: Yes you're right. I was having difficulties coming up with a good test. I'll come up with a good test. Thanks a lot!

Comment (piedcipher) [reply packages/flutter/test/material/expansion_panel_test.dart:1644]: Added a better widget-test 🤞. Please have a look.

File: packages/flutter/lib/src/material/expansion_panel.dart
Comment (piedcipher) [reply packages/flutter/lib/src/material/expansion_panel.dart]: Done 👍✅ Please have a look.

**Comment by piedcipher** [comment]
I've resolved review comments. Could you please review it again?

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (werainkhatri) [inline packages/flutter/test/material/expansion_panel_test.dart:1689]:
```suggestion
    expect(mergeableMaterial2.children.whereType<MaterialGap>().length, 1);
    expect(mergeableMaterial2.children.whereType<MaterialSlice>().length, 2);
```

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (werainkhatri) [inline packages/flutter/test/material/expansion_panel_test.dart:1701]:
```suggestion
    expect(mergeableMaterial3.children.whereType<MaterialGap>().length, 1);
    expect(mergeableMaterial3.children.whereType<MaterialSlice>().length, 2);
```

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (werainkhatri) [reply packages/flutter/test/material/expansion_panel_test.dart:1644]: oh this is great! thanks.

**Review by werainkhatri — APPROVED** [review]
thanks for another great contribution!

literal stamp

File: packages/flutter/test/material/expansion_panel_test.dart
Comment (piedcipher) [reply packages/flutter/test/material/expansion_panel_test.dart:1644]: I was unsure whether this new test was good or not.

**Review by piedcipher — COMMENTED** [review]
oh I didn't realise this mistake! Thanks for fixing

**Comment by piedcipher** [comment]
@werainkhatri google testing is still yellow.

**Comment by werainkhatri** [comment]
> @werainkhatri google testing is still yellow

it's fine. yellow is not bad or wrong, it's expected. anyway you have to wait for a second review. just be patient. we'll handle it and let you know if there is anything you need to do.

**Comment by piedcipher** [comment]
> > @werainkhatri google testing is still yellow
> 
> it's fine. yellow is not bad or wrong, it's expected. anyway you have to wait for a second review. just be patient. we'll handle it and let you know if there is anything you need to do.

No hurries, already keeping the patience (it's a month pr). Now I feel that you're pushing slightly hard for the rebase thing, I'll confirm it by asking a question in discord.

**Comment by piedcipher** [comment]
Sincere Apologies for the unnecessary rebases. I should have learnt this earlier. I didn't know what I was doing, I thought need to catch-up with main.

**Comment by piedcipher** [comment]
> > @werainkhatri google testing is still yellow
> 
> it's fine. yellow is not bad or wrong, it's expected. anyway you have to wait for a second review. just be patient. we'll handle it and let you know if there is anything you need to do.

Sure! _and_ I'm not in any hurry, it's a month old and I'm pretty patient and not in any hurry. Take your time.

**Review by Hixie — APPROVED** [review]
Thanks for the contribution!

LGTM

**Pull request merged**
